libc/string_catdup.c

```diff
@@ -0,0 +1,728 @@
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, SIZE_MAX));
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf);
+
+	for (i = strlen(WORD_1 WORD_2); i < SIZE_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+
+#pragma GCC diagnostic pop
+}
+
+
+TEST(string_cat, strncat_empty_dest)
+{
+	int i;
+	char buf[SIZE_BUF] = "";
+	char buf1[] = WORD_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, strlen(WORD_2) + 1));
+	TEST_ASSERT_EQUAL_STRING(WORD_2, buf);
+
+	for (i = strlen(WORD_2); i < SIZE_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strncat_empty_src)
+{
+	int i;
+	char buf[SIZE_BUF] = WORD_1;
+	char buf1[] = "";
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, 1));
+	TEST_ASSERT_EQUAL_STRING(WORD_1, buf);
+
+	for (i = strlen(WORD_1); i < SIZE_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strncat_huge_string_dest)
+{
+	char buf[PATH_MAX] = "";
+	char buf1[] = "";
+
+	memcpy(buf, hugeTestStr, PATH_MAX - 1);
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, 1));
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf);
+}
+
+
+TEST(string_cat, strncat_huge_string_src)
+{
+	char buf[PATH_MAX + 1] = "";
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, huge_buf, PATH_MAX));
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf);
+}
+
+
+TEST(string_cat, strncat_huge_string_both)
+{
+	char buf[PATH_MAX] = "";
+	char buf1[PATH_MAX / 2] = "";
+
+	memcpy(buf, huge_buf, PATH_MAX / 2);
+	memcpy(buf1, &huge_buf[PATH_MAX / 2], PATH_MAX / 2 - 1);
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, PATH_MAX / 2));
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf);
+}
+
+
+TEST(string_cat, strncat_specials_string)
+{
+	int i;
+	char buf[SIZE_BUF * 4] = SPECIALS_STRING;
+	char buf1[] = SPECIALS_STRING;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, strlen(SPECIALS_STRING SPECIALS_STRING)));
+	TEST_ASSERT_EQUAL_STRING(SPECIALS_STRING SPECIALS_STRING, buf);
+
+	for (i = strlen(SPECIALS_STRING SPECIALS_STRING); i < SIZE_BUF * 4; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+TEST(string_cat, strncat_ascii)
+{
+	int i;
+	char buf[SIZE_BIGGER_BUF] = { 0 };
+	char ascii[128] = { 0 };
+
+	for (i = 1; i < 128; i++) {
+		ascii[i - 1] = i;
+	}
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, ascii, i + 1));
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+	for (i = strlen(ascii); i < SIZE_BIGGER_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strncat_extended_ascii)
+{
+	int i;
+	char buf[SIZE_BIGGER_BUF] = { 0 };
+	char ascii[128] = { 0 };
+
+	for (i = 128; i < 255; i++) {
+		ascii[i - 128] = i;
+	}
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, ascii, i - 127));
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+	for (i = strlen(ascii); i < SIZE_BIGGER_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strncat_non_ascii)
+{
+	int i;
+	char buf[SIZE_BUF * 2] = NON_ASCII_1;
+	char buf1[] = NON_ASCII_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, strlen(NON_ASCII_1 NON_ASCII_2)));
```

Comment:
:exclamation: 
this is strange, almost none of the tests above check the most important thing about `strncat` - trimming src data to `n < strlen(src)` and always appending `\0` - `n` is always `0` or`  n >= strlen(src)`. 

I really don't care about contents of src (I can't fathom implementation in which it does) - but **I care deeply about the correct semantics in response to `n` value**

Reply:
We were considering passing the length of the source string by default (instead of len +1), but then we would have to ignore "-Wstringop-truncation" each time. We left one case for it and I hope it will be sufficient.

Reply:
We added a few cases that check it. We were considering more tests with truncation, but this utility seems to be covered with those tests.